(#634) restore custom domain when recreating project

angelorendina opened the pull request:
feat: restore previous custom domain (if any) when recreating a project (#634)

angelorendina commented on gateway/src/service.rs:
```diff
@@ -515,6 +526,26 @@ impl GatewayService {
             .map_err(|_| Error::from_kind(ErrorKind::Internal))
     }
 
+    pub async fn find_custom_domain_for_project(
```
Utility method for `project->domain` lookup. Inverse of the method below which performs `domain->project`.

angelorendina commented on gateway/src/service.rs:
```diff
@@ -428,7 +428,18 @@ impl GatewayService {
             let project = row.get::<SqlxJson<Project>, _>("project_state").0;
             if project.is_destroyed() {
                 // But is in `::Destroyed` state, recreate it
-                let project = Project::create(project_name.clone());
```
`Project::create` does not allow to customise the `ProjectCreating` payload, so manually creating the latter using the existing builder pattern. Perhaps there could be an ad-hoc builder for Project itself.

angelorendina commented on gateway/src/service.rs:
```diff
@@ -835,4 +866,51 @@ pub mod tests {
+            .unwrap();
+
+        assert!(matches!(recreated_project, Project::Creating(_)));
+        // TODO: check `recreated_project->fqdn == domain`
```
Possibly related to the issue above on how to instantiate a project. There is no way to check that the `fqdn` field in ProjectCreating is as expected since it's private. Could add a `#[cfg(test)]` getter but I haven't seen it done elsewhere in the project,

chesedo replied:
Making the field `pub(crate)` will also be acceptable

angelorendina replied:
I ended up making it a `cfg(test)` getter, I reckon it's more suggestive of its expected usage (and also ends up behaving as `pub(crate)` anyway).

oddgrd replied:
That makes sense, it shouldn't be more public than it needs to be. We can always make it pub(crate) in the future if we need it outside of tests. :ok_hand:

angelorendina commented on gateway/src/project.rs:
```diff
@@ -201,10 +201,6 @@ impl Project {
         }
     }
 
-    pub fn create(project_name: ProjectName) -> Self {
-        Self::Creating(ProjectCreating::new_with_random_initial_key(project_name))
-    }
-
```
(3rd commit proposal, can be removed if not liked)
Removing this public method as there is already a builder pattern for the inner `ProjectCreating`.
Alternatively one could change this method to accept a `fqdn: Option<String>` as well, to reflect the current usage of the method - I prefer the former though.

angelorendina reviewed (commented):
Rebased and pushed more commits, also waiting to see if the CI is happy this time so still keeping this a draft

oddgrd reviewed (approved):
LGTM, thanks Angelo!

angelorendina commented:
@oddgrd thanks for the review. I have just rebased from latest master and squashed the commits into one, so it should be good to go as per your guidelines 👍

chesedo commented on gateway/src/project.rs:
```diff
@@ -615,6 +611,11 @@ impl ProjectCreating {
         &self.initial_key
     }
 
+    #[cfg(test)]
+    pub fn fqdn(&self) -> &Option<String> {
+        &self.fqdn
+    }
+
```
```suggestion
    pub fn fqdn(&self) -> &Option<String> {
        &self.fqdn
    }

```

chesedo commented:
Thanks for this @angelorendina!
